staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go

```diff
@@ -0,0 +1,211 @@
+			// Return early
+			return obj, nil
+		}
+
+		// Convert CSA manager into SSA manager
+		managedFields[csaManagerIndex].Operation = metav1.ManagedFieldsOperationApply
+		managedFields[csaManagerIndex].Manager = ssaManagerName
+
+		err = unionManagerIntoIndex(managedFields, csaManagerIndex, csaManagerName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Create version of managed fields which has no CSA managers with the given name
+	filteredManagers := filter(managedFields, func(entry metav1.ManagedFieldsEntry) bool {
+		return !(entry.Manager == csaManagerName &&
+			entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+			entry.Subresource == "")
+	})
+
+	copied := obj.DeepCopyObject()
+	copiedAccessor, err := meta.Accessor(copied)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get meta accessor for copied object: %w", err)
+	}
+	copiedAccessor.SetManagedFields(filteredManagers)
+
+	// Wipe out CSA annotation if it exists
+	annotations := copiedAccessor.GetAnnotations()
+	delete(annotations, csaAnnotationName)
+	copiedAccessor.SetAnnotations(annotations)
+
+	return copied, nil
+}
+
+// Locates an Update manager entry named `csaManagerName` with the same APIVersion
+// as the manager at the targetIndex. Unions both manager's fields together
+// into the manager specified by `targetIndex`. No other managers are modified.
+func unionManagerIntoIndex(entries []metav1.ManagedFieldsEntry, targetIndex int, csaManagerName string) error {
+	ssaManager := entries[targetIndex]
+
+	// find Update manager of same APIVersion, union ssa fields with it.
+	// discard all other Update managers of the same name
+	csaManagerIndex, csaManagerExists := findFirstIndex(entries,
+		func(entry metav1.ManagedFieldsEntry) bool {
+			return entry.Manager == csaManagerName &&
+				entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+				entry.Subresource == "" &&
+				entry.APIVersion == ssaManager.APIVersion
+		})
+
+	targetFieldSet, err := fieldsToSet(*ssaManager.FieldsV1)
+	if err != nil {
+		return fmt.Errorf("failed to convert fields to set: %w", err)
+	}
+
+	combinedFieldSet := &targetFieldSet
+
+	// Union the csa manager with the existing SSA manager. Do nothing if
+	// there was no good candidate found
+	if csaManagerExists {
+		csaManager := entries[csaManagerIndex]
+
+		csaFieldSet, err := fieldsToSet(*csaManager.FieldsV1)
+		if err != nil {
+			return fmt.Errorf("failed to convert fields to set: %w", err)
+		}
+
+		combinedFieldSet = combinedFieldSet.Union(&csaFieldSet)
+	}
+
+	// Ensure that the resultant fieldset does not include the
+	// last applied annotation
+	combinedFieldSet = combinedFieldSet.Difference(csaAnnotationFieldSet)
+
+	// Encode the fields back to the serialized format
+	combinedFieldSetEncoded, err := setToFields(*combinedFieldSet)
+	if err != nil {
+		return fmt.Errorf("failed to encode field set: %w", err)
+	}
+
+	entries[targetIndex].FieldsV1 = &combinedFieldSetEncoded
```

Comment:
Probably same thing as above, we shouldn't force this to be v1, just in case we have more than 1 later.

Reply:
done